### PR TITLE
Improve image attachment management

### PR DESCRIPTION
**Pull request by eneiluj**
refs #2411

Implementation of the behaviour suggested by @max-nextcloud in https://github.com/nextcloud/text/issues/2411#issuecomment-1128776961 with which I completely agree :heart: :

* [x] 1. If the image url matches `.attachments.MD_FILE_ID/FILENAME` use FILENAME to get the image with the attachment API.
* [ ] 2. If the image url matches `.attachments.OTHER_FILE_ID/FILENAME`
  * [x] a. try to retrieve the image with the WebDav API.
  * [x] b. If that fails try to retrieve FILENAME with the attachment API.
  * [ ] c. Maybe rewrite the url with the own file id if b. succeeds.
* [x] 3. If not, use the path with the WebDav API.
* [x] 4. If that fails check if the url has a fileID param and try to get the file from that. (So we keep the old behavior for old images)

So this PR brings back the relative path in the markdown content's image target. We now write:
```
![IMAGE_FILE_NAME](.attachments.MARKDOWN_FILE_ID/ENCODED_IMAGE_FILE_NAME)
```

The problem with 2.a. and 2.b. is that `imageUrl` is a computed prop in which we would need a non-async (sync?) image loading check. But we can only make async network calls in a computed prop :grin:.
There already is a check which depends on `imageUrl`.
In short, we would need to make deeper changes to ImageView to allow 2.a. and 2.b.

We could discuss that and in the meantime I might come up with a first solution.

### CODE REVIEW

**Comment by eneiluj** [comment]
@max-nextcloud I found a reasonable solution for 2.a. and 2.b. : move the content of the `imageUrl` computed prop in a method (called `init`) which is called in `beforeMount` (which is the only time and place we loaded the image anyway). This way we can happily call async stuff.

`imageUrl` is now a data attribute. It is set if the image has been successfully loaded.

The image loading has been moved in a method as well (`loadImage`). It has an optional error callback that we use for 2.

The `init` method now handles all the different cases and is more flexible than the old `imageUrl` computed prop.

File: src/nodes/ImageView.vue [hunk 1]
```diff
@@ -301,6 +310,29 @@ export default {
...
+			} else {
+				return generateUrl('/apps/text/image?documentId={documentId}&sessionId={sessionId}&sessionToken={sessionToken}&imageFileName={imageFileName}&shareToken={shareToken}',
+					{
+						documentId,
+						sessionId,
+						sessionToken,
+						imageFileName,
+						shareToken: this.token,
+					})
+			}
```
Comment (vinicius73) [inline src/nodes/ImageView.vue:334]:
```suggestion
			}
		return generateUrl('/apps/text/image?documentId={documentId}&sessionId={sessionId}&sessionToken={sessionToken}&imageFileName={imageFileName}&shareToken={shareToken}',
			{
				documentId,
				sessionId,
				sessionToken,
				imageFileName,
				shareToken: this.token,
			})

```

File: src/nodes/ImageView.vue [hunk 2]
```diff
@@ -280,18 +248,59 @@ export default {
...
+				const imageFileName = getQueryVariable(this.src, 'imageFileName')
+				this.loadImage(this.getTextApiUrl(imageFileName))
+				return
+			}
+			if (this.src.startsWith(`.attachments.${this.currentSession?.documentId}/`)) {
+				const imageFileName = decodeURIComponent(this.src.replace(`.attachments.${this.currentSession?.documentId}/`, '').split('?')[0])
+				this.loadImage(this.getTextApiUrl(imageFileName))
+				return
+			}
+			if (this.isRemoteUrl || this.isPreviewUrl || this.isDataUrl) {
```
Comment (vinicius73) [inline src/nodes/ImageView.vue:265]: Can we turn it into a computed?
```suggestion
			if (this.isDirectUrl) {
```

File: src/nodes/ImageView.vue (hunk 1 above)
Comment (eneiluj) [reply src/nodes/ImageView.vue]: Why not.

File: src/nodes/ImageView.vue [hunk 3]
```diff
@@ -280,18 +248,59 @@ export default {
...
+				if (failureCallback !== null) {
+					failureCallback()
+				} else {
+					this.failed = true
+					this.imageLoaded = false
+					this.loaded = true
+				}
+			}
+			img.src = imageUrl
+		},
```
[the remark this answers is not in the document]
Comment (eneiluj) [reply src/nodes/ImageView.vue]: Thank you for showing me the way to the promises. Much better indeed :heart:

File: src/nodes/ImageView.vue (hunk 2 above)
Comment (eneiluj) [reply src/nodes/ImageView.vue]: Yep!

**Comment by eneiluj** [comment]
@vinicius73 Thanks for the review. If you have any idea about how to fix the Jest tests. It seems the checks are done before the async stuff is finished.

**Comment by eneiluj** [comment]
Let's commit the compiled code in the end to make our lives easier.

So don't forget to compile locally :wink:.

File: src/nodes/ImageView.vue [hunk 4]
```diff
@@ -280,18 +251,72 @@ export default {
...
-		img.onerror = () => {
+		this.init()
+	},
+	methods: {
+		init() {
+			if (this.src.startsWith('text://')) {
+				const imageFileName = getQueryVariable(this.src, 'imageFileName')
+				this.loadImage(this.getTextApiUrl(imageFileName)).catch((e) => {
+					this.onImageLoadFailure()
+				})
```
Comment (vinicius73) [inline src/nodes/ImageView.vue:262]: You can simplify this code.

```suggestion
				this.loadImage(this.getTextApiUrl(imageFileName))
				  .catch(this.onImageLoadFailure)
```

Comment (vinicius73) [reply src/nodes/ImageView.vue:262]: Also, we can use async/await and wrap `this.init()` and handle error from `this.init()`

```suggestion
				await this.loadImage(this.getTextApiUrl(imageFileName))
```

or

```suggestion
				return this.loadImage(this.getTextApiUrl(imageFileName))
```

and, change `this.init()` call.

```js
this.init().catch(this.onImageLoadFailure)
```

**Comment by max-nextcloud** [comment]
Thanks a lot for picking up the idea. :heart_eyes: 
I only had a quick look at the code and checked if the `text://` urls are still parsed and will still display nicely. Looks good to me in that regard.

**Comment by eneiluj** [comment]
@vinicius73 @max-nextcloud Ready to merge!

**Review by vinicius73 — APPROVED** [review]

File: tests/TextTest.php [hunk 5]
```diff
@@ -7,36 +7,54 @@
...
-			'aaa (2.png',
-			'aaa ((2.png',
-			'aaa 2)).png',
-			'a[a]a.png',
-			'a(a)a.png',
-			'a](a.png',
-			',;:!?.§-_a_',
-			'a`a`.png',
-		];
+	public function testGetOldAttachmentNamesFromContent() {
```
Comment (juliushaertl) [inline tests/TextTest.php:29]: Something for a separate PR but we should move those tests into tests/unit/Service/ImageServiceTest.php

File: lib/Service/ImageService.php [hunk 6]
```diff
@@ -476,20 +479,33 @@ public function cleanupAttachments(int $fileId): int {
...
+			$oldMatches,
+			PREG_SET_ORDER
+		);
+		$oldNames = array_map(static function (array $match) {
+			return urldecode($match[1]);
+		}, $oldMatches);
+
+		$matches = [];
+		preg_match_all(
+			'/\!\[(?>[^\[\]]+|\[(?>[^\[\]]+|\[(?>[^\[\]]+|\[(?>[^\[\]]+|\[(?>[^\[\]]+|\[(?>[^\[\]]+|\[\])*\])*\])*\])*\])*\])*\]\(\.attachments\.'.$fileId.'\/([^)&]+)\)/',
```
Comment (juliushaertl) [inline lib/Service/ImageService.php:499]: Would be nice to have a comment on what the regex should match to make is easier to catch its impact on the surrounding logic

**Review by juliushaertl — APPROVED** [review]
Minor comments, but feel free to address them separately :)

**Comment by eneiluj** [comment]
Should we backport this to stable24? I think we should.
